src/Microsoft.ML.Core/Data/WrappingRow.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+namespace Microsoft.ML.Runtime.Data
+{
+    /// <summary>
+    /// Convenient base class for <see cref="Row"/> implementors that wrap a single <see cref="Row"/>
+    /// as their input. The <see cref="Row.Position"/>, <see cref="Row.Batch"/>, and <see cref="Row.GetIdGetter"/>
+    /// are taken from this <see cref="Input"/>.
+    /// </summary>
+    [BestFriend]
+    internal abstract class WrappingRow : Row
+    {
+        private bool _disposed;
+
+        /// <summary>
+        /// The wrapped input row.
+        /// </summary>
+        protected Row Input { get; }
+
+        public sealed override long Batch => Input.Batch;
+        public sealed override long Position => Input.Position;
+        public override ValueGetter<UInt128> GetIdGetter() => Input.GetIdGetter();
+
+        [BestFriend]
+        private protected WrappingRow(Row input)
+        {
+            Contracts.AssertValue(input);
+            Input = input;
+        }
+
+        /// <summary>
+        /// This override of the dispose method by default only calls <see cref="Input"/>'s
+        /// <see cref="IDisposable.Dispose"/> method, but subclasses can enable additional functionality
+        /// via the <see cref="DisposeCore(bool)"/> functionality.
+        /// </summary>
+        /// <param name="disposing"></param>
+        protected sealed override void Dispose(bool disposing)
```

Comment:
[https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/dispose-pattern](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/dispose-pattern)
I think this is MS recommendation.

---
In reply to: [239561491](https://github.com/dotnet/machinelearning/pull/1835#discussion_r239561491) [](ancestors = 239561491)

Reply:
Yes, as @Ivanidzo4ka says, `disposing` is preferred. Also `Dispose(bool)` is already taken. I only diverged from the pattern sparingly, since disposal/finalization is one of those things that hard enough to get "right" that slavish devotion to the pattern really helps keep the code maintainable and correct, right down to their recommendations to the names of the parameters, the order of the `if` statements, order of `base` calls, and so on, and so on.